Compilation fails with Tpetra and Kokkos with complex and float

Trying to build the current master branch with `gcc 9.4.0` and `Trilinos 15.1.1` configured with `Trilinos_ENABLE_COMPLEX=ON` and `Trilinos_ENABLE_FLOAT=ON` on our house cluster, but I get a 1200 lines long error.

Does this look familiar to anyone? It rather looks like an issue in Trilinos/Kokkos than in deal.II.

```
FAILED: source/lac/CMakeFiles/object_lac_debug.dir/trilinos_tpetra_precondition.cc.o 
/linux-ubuntu20.04-skylake_avx512/gcc9.4.0/openmpi/4.1.6/k6qlwqr2vx5osaqf/bin/mpicxx -DDEBUG -DKOKKOS_DEPENDENCE -I/build-dealii/source/lac -I/dealii/source/lac -I/build-dealii/include -I/dealii/include -isystem /dealii/bundled/boost-1.84.0/include -isystem /dealii/bundled/umfpack/UMFPACK/Include -isystem /dealii/bundled/umfpack/AMD/Include -isystem /dealii/bundled/muparser_v2_3_3/include -isystem /linux-ubuntu20.04-skylake_avx512/gcc9.4.0/zlib/1.2.13/sz4h6ncyn4o7aczi/include -isystem /bin/trilinos-15.1.1/include -isystem /bin/sundials-7.0.0/include -isystem /bin/p4est-2.3.6/FAST/include -std=c++17 -fPIC -pedantic -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-placement-new -Wno-deprecated-declarations -Wno-literal-suffix -Wno-psabi -Wno-unused-local-typedefs -Wno-parentheses -fopenmp-simd -march=skylake-avx512 -O0 -ggdb -Wa,--compress-debug-sections -g -MD -MT source/lac/CMakeFiles/object_lac_debug.dir/trilinos_tpetra_precondition.cc.o -MF source/lac/CMakeFiles/object_lac_debug.dir/trilinos_tpetra_precondition.cc.o.d -o source/lac/CMakeFiles/object_lac_debug.dir/trilinos_tpetra_precondition.cc.o -c /dealii/source/lac/trilinos_tpetra_precondition.cc
In file included from /bin/trilinos-15.1.1/include/Tpetra_Map_decl.hpp:53,
                 from /bin/trilinos-15.1.1/include/Tpetra_Map.hpp:1,
                 from /dealii/include/deal.II/base/index_set.h:34,
                 from /dealii/include/deal.II/lac/trilinos_tpetra_precondition.templates.h:22,
                 from /dealii/source/lac/trilinos_tpetra_precondition.cc:16:
/bin/trilinos-15.1.1/include/Kokkos_DualView.hpp: In instantiation of 'Kokkos::DualView<DataType, Arg1Type, Arg2Type, Arg3Type>::DualView(const Kokkos::DualView<DT, DP ...>&) [with DT = std::complex<float>**; DP = {Kokkos::LayoutLeft, Kokkos::Serial, Kokkos::MemoryTraits<1>}; DataType = Kokkos::complex<float>**; Arg1Type = Kokkos::LayoutLeft; Arg2Type = Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>; Arg3Type = void]':
/dealii/include/deal.II/lac/trilinos_tpetra_precondition.templates.h:175:52:   required from 'void dealii::LinearAlgebra::TpetraWrappers::PreconditionBase<Number, MemorySpace>::vmult(dealii::Vector<OtherNumber>&, dealii::Vector<OtherNumber>&) const [with Number = std::complex<float>; MemorySpace = dealii::MemorySpace::Host]'
/dealii/source/lac/trilinos_tpetra_precondition.cc:66:20:   required from here
/bin/trilinos-15.1.1/include/Kokkos_DualView.hpp:293:26: error: no matching function for call to 'Kokkos::View<Kokkos::complex<float>**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void>::View(const t_dev&)'
  293 |         h_view(src.h_view) {}
      |                          ^
In file included from /bin/trilinos-15.1.1/include/Kokkos_Parallel.hpp:31,
                 from /bin/trilinos-15.1.1/include/Kokkos_MemoryPool.hpp:26,
                 from /bin/trilinos-15.1.1/include/Kokkos_TaskScheduler.hpp:34,
                 from /bin/trilinos-15.1.1/include/Serial/Kokkos_Serial.hpp:37,
                 from /bin/trilinos-15.1.1/include/decl/Kokkos_Declare_SERIAL.hpp:21,
                 from /bin/trilinos-15.1.1/include/KokkosCore_Config_DeclareBackend.hpp:22,
                 from /bin/trilinos-15.1.1/include/Kokkos_Core.hpp:45,
                 from /bin/trilinos-15.1.1/include/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp:4,
                 from /bin/trilinos-15.1.1/include/Tpetra_Details_DefaultTypes.hpp:45,
                 from /bin/trilinos-15.1.1/include/Tpetra_ConfigDefs.hpp:43,
                 from /bin/trilinos-15.1.1/include/Tpetra_Map_decl.hpp:47,
                 from /bin/trilinos-15.1.1/include/Tpetra_Map.hpp:1,
                 from /dealii/include/deal.II/base/index_set.h:34,
                 from /dealii/include/deal.II/lac/trilinos_tpetra_precondition.templates.h:22,
                 from /dealii/source/lac/trilinos_tpetra_precondition.cc:16:
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1640:35: note: candidate: 'Kokkos::View<DataType, Properties>::View(const typename Kokkos::View<DataType, Properties>::traits::execution_space::scratch_memory_space&, size_t, size_t, size_t, size_t, size_t, size_t, size_t, size_t) [with DataType = Kokkos::complex<float>**; Properties = {Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void}; typename Kokkos::View<DataType, Properties>::traits::execution_space::scratch_memory_space = Kokkos::ScratchMemorySpace<Kokkos::Serial>; size_t = long unsigned int]'
 1640 |   explicit KOKKOS_INLINE_FUNCTION View(
      |                                   ^~~~
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1641:69: note:   no known conversion for argument 1 from 'const t_dev' {aka 'const Kokkos::View<std::complex<float>**, Kokkos::LayoutLeft, Kokkos::Serial, Kokkos::MemoryTraits<1> >'} to 'const scratch_memory_space&' {aka 'const Kokkos::ScratchMemorySpace<Kokkos::Serial>&'}
 1641 |       const typename traits::execution_space::scratch_memory_space& arg_space,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1632:35: note: candidate: 'Kokkos::View<DataType, Properties>::View(const typename Kokkos::View<DataType, Properties>::traits::execution_space::scratch_memory_space&, const typename Kokkos::View<DataType, Properties>::traits::array_layout&) [with DataType = Kokkos::complex<float>**; Properties = {Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void}; typename Kokkos::View<DataType, Properties>::traits::execution_space::scratch_memory_space = Kokkos::ScratchMemorySpace<Kokkos::Serial>; typename Kokkos::View<DataType, Properties>::traits::array_layout = Kokkos::LayoutLeft]'
 1632 |   explicit KOKKOS_INLINE_FUNCTION View(
      |                                   ^~~~
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1632:35: note:   candidate expects 2 arguments, 1 provided
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1585:35: note: candidate: 'Kokkos::View<DataType, Properties>::View(Kokkos::View<DataType, Properties>::pointer_type, const typename Kokkos::View<DataType, Properties>::traits::array_layout&) [with DataType = Kokkos::complex<float>**; Properties = {Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void}; Kokkos::View<DataType, Properties>::pointer_type = Kokkos::complex<float>*; typename Kokkos::View<DataType, Properties>::traits::array_layout = Kokkos::LayoutLeft]'
 1585 |   explicit KOKKOS_INLINE_FUNCTION View(
      |                                   ^~~~
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1585:35: note:   candidate expects 2 arguments, 1 provided
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1568:35: note: candidate: 'Kokkos::View<DataType, Properties>::View(Kokkos::View<DataType, Properties>::pointer_type, size_t, size_t, size_t, size_t, size_t, size_t, size_t, size_t) [with DataType = Kokkos::complex<float>**; Properties = {Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void}; Kokkos::View<DataType, Properties>::pointer_type = Kokkos::complex<float>*; size_t = long unsigned int]'
 1568 |   explicit KOKKOS_INLINE_FUNCTION View(
      |                                   ^~~~
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1569:20: note:   no known conversion for argument 1 from 'const t_dev' {aka 'const Kokkos::View<std::complex<float>**, Kokkos::LayoutLeft, Kokkos::Serial, Kokkos::MemoryTraits<1> >'} to 'Kokkos::View<Kokkos::complex<float>**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void>::pointer_type' {aka 'Kokkos::complex<float>*'}
 1569 |       pointer_type arg_ptr, const size_t arg_N0 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
      |       ~~~~~~~~~~~~~^~~~~~~
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1539:26: note: candidate: 'template<class Traits> Kokkos::View<DataType, Properties>::View(const typename Kokkos::View<DataType, Properties>::view_tracker_type::track_type&, const Kokkos::Impl::ViewMapping<Traits, typename Traits::specialize>&)'
 1539 |   KOKKOS_INLINE_FUNCTION View(
      |                          ^~~~
/bin/trilinos-15.1.1/include/Kokkos_View.hpp:1539:26: note:   template argument deduction/substitution failed:
[...]
```

[full error message](https://github.com/user-attachments/files/16495524/full-error.txt)

--- a/examples/step-86/step-86.cc
+++ b/examples/step-86/step-86.cc
@@ -591,7 +591,7 @@ namespace Step86
     // the constrained dofs corresponding to hanging nodes (i.e., those for
     // which the lines of the `current_constraints` contain at least one other
     // entry), and to the difference between the input vector and the actual
-    // solution on those constraints that correspond to boundary conditions. 
+    // solution on those constraints that correspond to boundary conditions.
     for (const auto &c : current_constraints.get_lines())
       if (locally_owned_dofs.is_element(c.index))
         {

--- a/include/deal.II/lac/trilinos_tpetra_precondition.templates.h
+++ b/include/deal.II/lac/trilinos_tpetra_precondition.templates.h
@@ -151,12 +151,18 @@ namespace LinearAlgebra
 
       // 1. Make host view with underlying dealii::Vector data.
       //    (This is always a n x 1 matrix since multivector expects that.)
-      TpetraTypes::HostViewTypeUnmanaged<Number> host_src(src.data(),
-                                                          src.size(),
-                                                          1);
-      TpetraTypes::HostViewTypeUnmanaged<Number> host_dst(dst.data(),
-                                                          dst.size(),
-                                                          1);
+
+      // Tpetra uses Kokkos::complex<> internally for std::complex<>
+      using value_type = typename TpetraTypes::HostViewType<Number>::value_type;
+      static_assert((numbers::NumberTraits<Number>::is_complex &&
+                     sizeof(value_type) == sizeof(Number)) ||
+                    (!numbers::NumberTraits<Number>::is_complex &&
+                     std::is_same_v<value_type, Number>));
+
+      TpetraTypes::HostViewType<Number> host_src(
+        reinterpret_cast<value_type *>(src.data()), src.size(), 1);
+      TpetraTypes::HostViewType<Number> host_dst(
+        reinterpret_cast<value_type *>(dst.data()), dst.size(), 1);
 
       // 2. Create device mirror (if Device == Host, no alloc )
       auto dev_src = Kokkos::create_mirror_view_and_copy(
@@ -166,10 +172,10 @@ namespace LinearAlgebra
         typename MemorySpace::kokkos_space::execution_space(), host_dst);
 
       // 3. Create dual views from the previous ones
-      TpetraTypes::DualViewTypeUnmanaged<Number, MemorySpace> dual_src(
-        dev_src, host_src);
-      TpetraTypes::DualViewTypeUnmanaged<Number, MemorySpace> dual_dst(
-        dev_dst, host_dst);
+      TpetraTypes::DualViewType<Number, MemorySpace> dual_src(dev_src,
+                                                              host_src);
+      TpetraTypes::DualViewType<Number, MemorySpace> dual_dst(dev_dst,
+                                                              host_dst);
 
       // 4. Create Tpetra Vector from
       TpetraTypes::VectorType<Number, MemorySpace> tpetra_dst(
@@ -201,12 +207,18 @@ namespace LinearAlgebra
 
       // 1. Make host view with underlying dealii::Vector data.
       //    (This is always a n x 1 matrix since multivector expects that.)
-      TpetraTypes::HostViewTypeUnmanaged<Number> host_src(src.data(),
-                                                          src.size(),
-                                                          1);
-      TpetraTypes::HostViewTypeUnmanaged<Number> host_dst(dst.data(),
-                                                          dst.size(),
-                                                          1);
+
+      // Tpetra uses Kokkos::complex<> internally for std::complex<>
+      using value_type = typename TpetraTypes::HostViewType<Number>::value_type;
+      static_assert((numbers::NumberTraits<Number>::is_complex &&
+                     sizeof(value_type) == sizeof(Number)) ||
+                    (!numbers::NumberTraits<Number>::is_complex &&
+                     std::is_same_v<value_type, Number>));
+
+      TpetraTypes::HostViewType<Number> host_src(
+        reinterpret_cast<value_type *>(src.data()), src.size(), 1);
+      TpetraTypes::HostViewType<Number> host_dst(
+        reinterpret_cast<value_type *>(dst.data()), dst.size(), 1);
 
       // 2. Create device mirror (if Device == Host, no alloc )
       auto dev_src = Kokkos::create_mirror_view_and_copy(
@@ -216,10 +228,10 @@ namespace LinearAlgebra
         typename MemorySpace::kokkos_space::execution_space(), host_dst);
 
       // 3. Create dual views from the previous ones
-      TpetraTypes::DualViewTypeUnmanaged<Number, MemorySpace> dual_src(
-        dev_src, host_src);
-      TpetraTypes::DualViewTypeUnmanaged<Number, MemorySpace> dual_dst(
-        dev_dst, host_dst);
+      TpetraTypes::DualViewType<Number, MemorySpace> dual_src(dev_src,
+                                                              host_src);
+      TpetraTypes::DualViewType<Number, MemorySpace> dual_dst(dev_dst,
+                                                              host_dst);
 
       // 4. Create Tpetra Vector from
       TpetraTypes::VectorType<Number, MemorySpace> tpetra_dst(

--- a/include/deal.II/lac/trilinos_tpetra_types.h
+++ b/include/deal.II/lac/trilinos_tpetra_types.h
@@ -145,17 +145,6 @@ namespace LinearAlgebra
       using RowMatrixType =
         Tpetra::RowMatrix<Number, LO, GO, NodeType<MemorySpace>>;
 
-
-      /**
-       * @brief Typedef for the Kokkos::View type.
-       * This is needed for shallow copies of deal.II LA structures
-       * to Trilinos LA structures.
-       *
-       */
-      template <typename Number>
-      using HostViewType =
-        typename VectorType<Number, dealii::MemorySpace::Host>::host_view_type;
-
       /**
        * @brief Typedef for the Kokkos::DualView type.
        * This is needed for shallow copies of deal.II LA structures
@@ -165,6 +154,15 @@ namespace LinearAlgebra
       using DualViewType =
         typename VectorType<Number, MemorySpace>::dual_view_type;
 
+      /**
+       * @brief Typedef for the Kokkos::View type.
+       * This is needed for shallow copies of deal.II LA structures
+       * to Trilinos LA structures.
+       *
+       */
+      template <typename Number>
+      using HostViewType =
+        typename DualViewType<Number, dealii::MemorySpace::Host>::t_host;
 
 #  ifdef DEAL_II_TRILINOS_WITH_IFPACK2
       /**

--- a/include/deal.II/lac/trilinos_tpetra_types.h
+++ b/include/deal.II/lac/trilinos_tpetra_types.h
@@ -21,6 +21,8 @@
 
 #ifdef DEAL_II_TRILINOS_WITH_TPETRA
 
+#  include <deal.II/base/memory_space.h>
+
 #  include <Kokkos_DualView.hpp>
 #  include <Tpetra_Details_DefaultTypes.hpp>
 
@@ -145,36 +147,23 @@ namespace LinearAlgebra
 
 
       /**
-       * @brief Typedef for the Kokkos::View type of unmanaged host memory.
-       *
-       * Unmanaged means that the view is not allocating data itself, but only
-       * uses an external pointer. Consequently, the memory is not freed upon
-       * destruction of the view.
+       * @brief Typedef for the Kokkos::View type.
        * This is needed for shallow copies of deal.II LA structures
        * to Trilinos LA structures.
        *
        */
       template <typename Number>
-      using HostViewTypeUnmanaged = Kokkos::View<Number **,
-                                                 Kokkos::LayoutLeft,
-                                                 Kokkos::HostSpace,
-                                                 Kokkos::MemoryUnmanaged>;
+      using HostViewType =
+        typename VectorType<Number, dealii::MemorySpace::Host>::host_view_type;
 
       /**
-       * @brief Typedef for the Kokkos::View type of unmanaged memory.
-       *
-       * Unmanaged means that the view is not allocating data itself, but only
-       * uses an external pointer. Consequently, the memory is not freed upon
-       * destruction of the view.
+       * @brief Typedef for the Kokkos::DualView type.
        * This is needed for shallow copies of deal.II LA structures
        * to Trilinos LA structures.
        */
       template <typename Number, typename MemorySpace>
-      using DualViewTypeUnmanaged =
-        Kokkos::DualView<Number **,
-                         Kokkos::LayoutLeft,
-                         typename MemorySpace::kokkos_space::execution_space,
-                         Kokkos::MemoryUnmanaged>;
+      using DualViewType =
+        typename VectorType<Number, MemorySpace>::dual_view_type;
 
 
 #  ifdef DEAL_II_TRILINOS_WITH_IFPACK2

